src/assets/Generator.Shared/ArmOperationHelpers.cs

```diff
@@ -0,0 +1,389 @@
+        /// <typeparam name="TResult">The response type</typeparam>
+        /// <param name="operation">The operation to wait upon</param>
+        /// <param name="cancellationToken">A cancellation token for the operation</param>
+        /// <returns></returns>
+        public static Response<TResult> WaitForCompletion<TResult>(this Operation<TResult> operation, CancellationToken cancellationToken = default) where TResult : notnull
+        {
+            return operation.WaitForCompletion(OperationHelpers.DefaultPollingInterval, cancellationToken);
+        }
+
+        /// <summary>
+        /// Waits for the long-running operation to complete, including a specified polling internal.
+        /// </summary>
+        /// <typeparam name="TResult">The response type</typeparam>
+        /// <param name="operation">The operation to wait upon</param>
+        /// <param name="pollingInterval">The duration to wait in-between each poll</param>
+        /// <param name="cancellationToken">A cancellation token for the operation</param>
+        /// <returns></returns>
+        public static Response<TResult> WaitForCompletion<TResult>(this Operation<TResult> operation, TimeSpan pollingInterval, CancellationToken cancellationToken = default) where TResult : notnull
+        {
+            while (true)
+            {
+                operation.UpdateStatus(cancellationToken);
+                if (operation.HasCompleted)
+                {
+                    return Response.FromValue(operation.Value, operation.GetRawResponse());
+                }
+
+                Thread.Sleep(pollingInterval);
+            }
+        }
+
+        internal static Operation<T> Create<T>(HttpPipeline pipeline, ClientDiagnostics clientDiagnostics, Response originalResponse, RequestMethod requestMethod, string scopeName, FinalStateVia finalStateVia,
+            Func<HttpMessage> createOriginalHttpMessage, Func<Response, CancellationToken, Response<T>> createFinalResponse, Func<Response, CancellationToken, ValueTask<Response<T>>> createFinalResponseAsync) where T : notnull
+        {
+            using HttpMessage originalHttpMethod = createOriginalHttpMessage();
+            string originalUri = originalHttpMethod.Request.Uri.ToString();
+            ScenarioInfo info = GetScenarioInfo(originalResponse, originalUri, requestMethod, finalStateVia);
+            if ((requestMethod != RequestMethod.Put && (info.HeaderFrom == HeaderFrom.None || !info.HasLocation)) || finalStateVia == FinalStateVia.AzureAsyncOperation)
+            {
+                // Support Operation Resource: https://github.com/Azure/autorest.csharp/issues/447
+                throw clientDiagnostics.CreateRequestFailedException(originalResponse);
+            }
+
+            return new ArmOperation<T>(pipeline, clientDiagnostics, originalResponse, scopeName, info, createFinalResponse, createFinalResponseAsync);
+        }
+
+        private class ArmOperation<T> : Operation<T> where T : notnull
```

Comment:
I know it's a lot to ask but can we summerize the logic in a comment above this type. It seems to be complicated enough to warrant that.

Reply:
Since I was going to do it anyway, I'm going to create a document for this information based on my diagram information. For now, I'm putting links to the reference documents I used to get to this point.